libraries/rush-lib/src/logic/PackageJsonUpdater.ts

```diff
@@ -151,6 +181,196 @@ export class PackageJsonUpdater {
+        commonVersionsConfiguration.preferredVersions.get(moduleName);
+
+      const version: string = await this._getNormalizedVersionSpec(
+        projects,
+        moduleName,
+        latestVersion,
+        implicitlyPreferredVersion,
+        explicitlyPreferredVersion,
+        inferredRangeStyle
+      );
+
+      if (devDependency) {
+        devDependenciesToUpdate[moduleName] = version;
+      } else {
+        dependenciesToUpdate[moduleName] = version;
+      }
+
+      console.log(colors.green(`Updating projects to use `) + moduleName + '@' + colors.cyan(version));
+      console.log();
+
+      const existingSpecifiedVersions: Set<string> | undefined = allVersionsByPackageName.get(moduleName);
+      if (
+        existingSpecifiedVersions &&
+        !existingSpecifiedVersions.has(version) &&
+        this._rushConfiguration.ensureConsistentVersions &&
+        !updateOtherPackages
+      ) {
+        // There are existing versions, and the version we're going to use is not one of them, and this repo
+        // requires consistent versions, and we aren't going to update other packages, so we can't proceed.
+
+        const existingVersionList: string = Array.from(existingSpecifiedVersions).join(', ');
+        throw new Error(
+          `Adding '${moduleName}@${version}' ` +
+            `causes mismatched dependencies. Use the "--make-consistent" flag to update other packages to use ` +
+            `this version, or try specify one of the existing versions (${existingVersionList}).`
+        );
+      }
+    }
+
+    const allPackageUpdates: IUpdateProjectOptions[] = [];
+
+    for (const project of projects) {
+      const currentProjectDepUpdate: IUpdateProjectOptions = {
+        project: new VersionMismatchFinderProject(project),
+        dependenciesToAddOrUpdateOrRemove: dependenciesToUpdate,
+        dependencyType: DependencyType.Regular
+      };
+
+      const currentProjectDevDepUpdate: IUpdateProjectOptions = {
+        project: new VersionMismatchFinderProject(project),
+        dependenciesToAddOrUpdateOrRemove: devDependenciesToUpdate,
+        dependencyType: DependencyType.Dev
+      };
+
+      this.updateProject(currentProjectDepUpdate);
+      this.updateProject(currentProjectDevDepUpdate);
+
+      const otherPackageUpdates: IUpdateProjectOptions[] = [];
+
+      if (this._rushConfiguration.ensureConsistentVersions || updateOtherPackages) {
+        // TODO: Reenable to support mismatch checks for both devDeps and regularDeps
+        // we need to do a mismatch check
+        const mismatchFinder: VersionMismatchFinder = VersionMismatchFinder.getMismatches(
+          this._rushConfiguration,
+          {
+            variant: variant
+          }
+        );
+        const mismatches: string[] = mismatchFinder.getMismatches().filter((mismatch) => {
+          console.log(mismatch);
+          return !projects.find((proj) => proj.packageName === mismatch);
```

Comment:
I realize performance here is not a huge deal, but since `packageName` is required to be unique within the workspace, you should create a filtered `Map<string, RushConfigurationProject>` first and do `.has()` checks on it for O(1) instead of O(n)